modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/downsampling/ReplaceSourceWithDownsampleIndexTask.java

```diff
@@ -0,0 +1,220 @@
+            if (dataStream != null
+                && dataStream.getIndices().stream().filter(index -> index.getName().equals(downsampleIndex)).findAny().isEmpty()) {
+                // add downsample index to data stream
+                LOGGER.trace(
+                    "unable find source index [{}] but adding index [{}] to data stream [{}]",
+                    sourceBackingIndex,
+                    downsampleIndex,
+                    dataStreamName
+                );
+                Metadata.Builder newMetaData = Metadata.builder(state.metadata())
+                    .put(dataStream.addBackingIndex(state.metadata(), downsampleIndexMeta.getIndex()));
+                return ClusterState.builder(state).metadata(newMetaData).build();
+            }
+        } else {
+            // the source index exists
+            DataStream sourceParentDataStream = sourceIndexAbstraction.getParentDataStream();
+            if (sourceParentDataStream != null) {
+                assert sourceParentDataStream.getName().equals(dataStreamName)
+                    : "the backing index must be part of the provided data "
+                        + "stream ["
+                        + dataStreamName
+                        + "] but it is instead part of data stream ["
+                        + sourceParentDataStream.getName()
+                        + "]";
+                if (sourceParentDataStream.getWriteIndex().getName().equals(sourceBackingIndex)) {
+                    String errorMessage = String.format(
+                        Locale.ROOT,
+                        "index [%s] is the write index for data stream [%s] and cannot be replaced",
+                        sourceBackingIndex,
+                        sourceParentDataStream.getName()
+                    );
+                    throw new IllegalStateException(errorMessage);
+                }
+                if (sourceIndexMeta != null) {
+                    // both indices exist, let's copy the origination date from the source index to the downsample index
+                    Metadata.Builder newMetaData = Metadata.builder(state.getMetadata());
+                    IndexMetadata updatedDownsampleMetadata = copyDataStreamLifecycleState(
+                        sourceIndexMeta,
+                        downsampleIndexMeta,
+                        sourceParentDataStream
+                    );
+
+                    newMetaData.put(updatedDownsampleMetadata, true);
+                    // replace source with downsample
+                    newMetaData.put(dataStream.replaceBackingIndex(sourceIndexMeta.getIndex(), downsampleIndexMeta.getIndex()));
+                    return ClusterState.builder(state).metadata(newMetaData).build();
+                }
+            } else {
+                // the source index is not part of a data stream, so let's check if we can make sure the downsample index ends up in the
+                // data stream
```

Comment:
Should we still try to copy stuff from the source index metadata if it exists, even if it's not in the data stream? (like the origination date?)

Reply:
Done